operator/cmd/mesh/uninstall.go

```diff
@@ -0,0 +1,182 @@
+}
+
+func UninstallCmd(logOpts *log.Options) *cobra.Command {
+	rootArgs := &rootArgs{}
+	uiArgs := &uninstallArgs{}
+	uicmd := &cobra.Command{
+		Use:   "uninstall --revision foo",
+		Short: "uninstall the control plane by revision",
+		Long:  "The uninstall command uninstall the control plane by revision and",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if uiArgs.revision == "" && uiArgs.filename == "" {
+				return fmt.Errorf("either one of the --revision or --filename flags needed to be set")
+			}
+			if len(args) > 0 {
+				return fmt.Errorf("too many positional arguments")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return uninstall(cmd, rootArgs, uiArgs, logOpts)
+		}}
+	addUninstallFlags(uicmd, uiArgs)
+	return uicmd
+}
+
+func uninstall(cmd *cobra.Command, rootArgs *rootArgs, uiArgs *uninstallArgs, logOpts *log.Options) error {
+	l := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), installerScope)
+	if err := configLogs(logOpts); err != nil {
+		return fmt.Errorf("could not configure logs: %s", err)
+	}
+	restConfig, _, clt, err := K8sConfig(uiArgs.kubeConfigPath, uiArgs.context)
+	if err != nil {
+		return err
+	}
+
+	if err := uninstallWrapper(cmd, rootArgs, uiArgs, restConfig, clt, l); err != nil {
+		return fmt.Errorf("failed to uninstall control plane: %v", err)
+	}
+	return nil
+}
+
+// uninstallWrapper is helper function to uninstalls control plane by either
+// pruning by target revision or deleting specified manifests.
+func uninstallWrapper(cmd *cobra.Command, rootArgs *rootArgs, uiArgs *uninstallArgs,
+	restConfig *rest.Config, clt client.Client, l clog.Logger) error {
+	cache.FlushObjectCaches()
+	opts := &helmreconciler.Options{DryRun: rootArgs.dryRun, Log: l, ProgressLog: progress.NewLog()}
+	h, err := helmreconciler.NewHelmReconciler(clt, restConfig, nil, opts)
+	if err != nil {
+		return fmt.Errorf("failed to create reconciler: %v", err)
+	}
+
+	if uiArgs.revision != "" {
+		objectsList, tp, err := h.GetPrunedResourcesList(uiArgs.revision)
+		if err != nil {
+			return err
+		}
+		if err := preCheckWarnings(cmd, rootArgs, uiArgs, uiArgs.revision, tp); err != nil {
+			return fmt.Errorf("failed to do preuninstall check: %v", err)
+		}
+
+		if err := h.PruneControlPlaneByRevision(uiArgs.revision, objectsList); err != nil {
+			return fmt.Errorf("failed to prune control plane resources by revision: %v", err)
+		}
+	} else if uiArgs.filename != "" {
+		manifestMap, iops, err := manifest.GenManifests([]string{uiArgs.filename}, applyFlagAliases(uiArgs.set, uiArgs.manifestsPath, uiArgs.revision), uiArgs.force, restConfig, l)
+		if err != nil {
+			return err
+		}
+		if iops == nil {
+			return fmt.Errorf("istioOperatorSpec is nil")
+		}
+		if err := preCheckWarnings(cmd, rootArgs, uiArgs, iops.Revision, nil); err != nil {
+			return fmt.Errorf("failed to do preuninstall check: %v", err)
```

Comment:
if preCheckWarnings doesn't report the exact error that it hits, it will not help the user to know something in preuninstall check failed. in that case the problem is in preCheckWarnings not issuing helpful errors. otoh if that function reports the exact check that failed, adding the fact that it's part of preuninstall check does not add much value imo.